File: src/lib/schemas.ts
import { z } from "zod";

export const navigateRequestSchema = z.object({
  country: z.literal("Germany"),
  concern: z.string().min(8, "Tell us a little more about the concern."),
  duration: z.string().min(1, "Choose a duration."),
  severity: z.coerce.number().min(1).max(10),
  symptoms: z.array(z.string()).default([]),
  languagePreference: z.enum(["english", "german_phrase"]).default("german_phrase"),
  userType: z
    .enum([
      "international_student",
      "tourist",
      "migrant_new_resident",
      "local_resident",
    ])
    .default("international_student"),
  tried: z.string().optional().default(""),
  severeConcern: z.boolean().optional().default(false),
});

export const navigationResultSchema = z.object({
  riskLevel: z.enum([
    "emergency",
    "urgent",
    "routine",
    "self_care_or_pharmacy",
    "administrative",
  ]),
  recommendedRoute: z.string(),
  reasoning: z.string(),
  urgentWarningSigns: z.array(z.string()),
  localTerms: z.array(z.object({ term: z.string(), meaning: z.string() })),
  documentsToBring: z.array(z.string()),
  doctorSummary: z.object({
    mainConcern: z.string(),
    duration: z.string(),
    severity: z.string(),
    keySymptoms: z.array(z.string()),
    whatToSay: z.string(),
    questionsToAsk: z.array(z.string()),
  }),
  localLanguagePhrase: z.object({
    language: z.string(),
    phrase: z.string(),
    meaning: z.string(),
  }),
  nextStep: z.string(),
  disclaimer: z.string(),
  source: z.enum(["rules", "gemini", "mock", "fastapi"]).default("mock"),
});

export const chatMessageSchema = z.object({
  role: z.enum(["user", "assistant"]),
  content: z.string().min(1).max(1600),
});

export const chatRequestSchema = z.object({
  situation: navigateRequestSchema,
  report: navigationResultSchema,
  messages: z.array(chatMessageSchema).min(1).max(12),
});

export const chatResponseSchema = z.object({
  reply: z.string(),
  source: z.enum(["rules", "gemini", "mock", "fastapi"]),
  urgent: z.boolean(),
});

File: backend/app/schemas.py
from typing import Literal

from pydantic import BaseModel, Field


RiskLevel = Literal["emergency", "urgent", "routine", "self_care_or_pharmacy", "administrative"]


class NavigateRequest(BaseModel):
    country: Literal["Germany"] = "Germany"
    concern: str = Field(min_length=8)
    duration: str
    severity: int = Field(ge=1, le=10)
    symptoms: list[str] = []
    languagePreference: Literal["english", "german_phrase"] = "german_phrase"
    userType: Literal["international_student", "tourist", "migrant_new_resident", "local_resident"] = "international_student"
    tried: str = ""
    severeConcern: bool = False


class LocalTerm(BaseModel):
    term: str
    meaning: str


class DoctorSummary(BaseModel):
    mainConcern: str
    duration: str
    severity: str
    keySymptoms: list[str]
    whatToSay: str
    questionsToAsk: list[str]


class LocalLanguagePhrase(BaseModel):
    language: str
    phrase: str
    meaning: str


class NavigationResult(BaseModel):
    riskLevel: RiskLevel
    recommendedRoute: str
    reasoning: str
    urgentWarningSigns: list[str]
    localTerms: list[LocalTerm]
    documentsToBring: list[str]
    doctorSummary: DoctorSummary
    localLanguagePhrase: LocalLanguagePhrase
    nextStep: str
    disclaimer: str
    source: Literal["rules", "gemini", "mock", "fastapi"] = "fastapi"


class ChatMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str = Field(min_length=1, max_length=1600)


class ChatRequest(BaseModel):
    situation: NavigateRequest
    report: NavigationResult
    messages: list[ChatMessage] = Field(min_length=1, max_length=12)


class ChatResponse(BaseModel):
    reply: str
    source: Literal["rules", "gemini", "mock", "fastapi"]
    urgent: bool
